File: lib/go/internal/RequiresAuthentication.go
//|
//|  Copyright The Telepact Authors
//|  SPDX-License-Identifier: Apache-2.0
//|

package internal

func RequiresAuthentication(schema SchemaAccessor, functionName string) bool {
	var functionDefinition map[string]any
	for _, definition := range schema.FullDefinitions() {
		definitionMap, ok := definition.(map[string]any)
		if !ok {
			continue
		}
		if _, exists := definitionMap[functionName]; !exists {
			continue
		}
		functionDefinition = definitionMap
		break
	}

	if functionName == "" || functionName[len(functionName)-1] == '_' {
		return false
	}
	if internal, ok := functionDefinition["$internal"].(bool); ok && internal {
		return false
	}
	if _, exists := schema.ParsedDefinitions()["union.Auth_"]; !exists {
		return false
	}
	if authenticated, ok := functionDefinition["$authenticated"].(bool); ok && !authenticated {
		return false
	}
	return true
}

File: lib/ts/src/internal/RequiresAuthentication.ts
//|
//|  Copyright The Telepact Authors
//|  SPDX-License-Identifier: Apache-2.0
//|

import { TelepactSchema } from '../TelepactSchema.js';

export function requiresAuthentication(telepactSchema: TelepactSchema, functionName: string): boolean {
    const functionDefinition = telepactSchema.full.find(
        (definition): definition is Record<string, unknown> =>
            typeof definition === 'object'
            && definition !== null
            && !Array.isArray(definition)
            && functionName in definition,
    );
    if (functionName.endsWith('_') || functionDefinition?.['$internal'] === true) {
        return false;
    }

    if (!('union.Auth_' in telepactSchema.parsed)) {
        return false;
    }

    return functionDefinition?.['$authenticated'] !== false;
}
